WordPress/Classes/ViewRelated/Domains/Domain registration/DomainPurchasingWebFlowController.swift

```diff
@@ -0,0 +1,338 @@
+        } else {
+            completion(false)
+        }
+    }
+
+    /// Handles URL changes in the web view.  We only allow the user to stay within certain URLs.  Falling outside these URLs
+    /// results in the web view being dismissed.  This method also handles the success condition for a successful domain registration
+    /// through said web view.
+    ///
+    /// - Parameters:
+    ///     - newURL: the newly set URL for the web view.
+    ///     - siteID: the ID of the site we're trying to register the domain against.
+    ///     - domain: the domain the user is purchasing.
+    ///     - completion: the closure that will be executed when the domain registration succeeds or fails.
+    ///
+    private func handleWebViewURLChange(
+        _ newURL: URL,
+        oldURL: URL?,
+        domain: Domain,
+        completion: (Result<String, DomainPurchasingError>) -> Void
+    ) {
+        let canOpenNewURL = newURL.absoluteString.starts(with: Constants.checkoutURLPrefix)
+        guard canOpenNewURL else {
+            let error = DomainPurchasingError.unsupportedRedirect(fromURL: oldURL, toURL: newURL)
+            completion(.failure(error))
+            return
+        }
+
+        let domainRegistrationSucceeded = newURL.absoluteString.starts(with: Constants.checkoutSuccessURLPrefix)
+
+        if domainRegistrationSucceeded {
+            completion(.success(domain.domainName))
+        }
+    }
+
+    /// Nullifies the variables that were set during a domain purchasing execution. In other words, it nullifies the variables under "Execution Variables" pragma mark.
+    private func cleanupExecutionVariables() {
+        self.webViewURLChangeObservation?.invalidate()
+        self.webViewURLChangeObservation = nil
+    }
+
+    /// Metadata to attach to error or track events.
+    private func userInfo(domain: DomainSuggestion, site: Blog) -> [String: Any] {
+        let homeURL = site.homeURL as String?
+        var userInfo = [String: Any]()
+        userInfo["siteID"] = site.dotComID?.intValue
+        userInfo["siteHomeURL"] = homeURL
+        userInfo["siteHostURL"] = URL(string: homeURL ?? "")?.host
+        userInfo["domainName"] = domain.domainName
+        userInfo["domainSupportsPrivacy"] = domain.supportsPrivacy
+        userInfo["checkoutWebAddress"] = Self.Constants.checkoutWebAddress
+        if let presentingViewController {
+            userInfo["presentingViewController"] = String(describing: type(of: presentingViewController))
+        }
+        return userInfo
+    }
+
+    /// Metadata to attach when capturing errors.
+    private func userInfoForError(_ error: DomainPurchasingError, domain: DomainSuggestion, site: Blog) -> [String: Any] {
+        var userInfo = userInfo(domain: domain, site: site)
+        userInfo = userInfo.merging(error.errorUserInfo) { $1 }
+        return userInfo
+    }
+
+    // MARK: - Types
+
+    typealias CompletionHandler = (Result<String, DomainPurchasingError>) -> Void
```

Comment:
This could be more descriptive I think. `DomainPurchasingWebFlowController.CompletionHandler` may not be very expressive.

Reply:
So the name `CompletionHandler` sounds pretty generic to me as most closures are used as completion closures. If you think that's straightforward enough, it is okay for me.